addons/godottpd/http_file_router.gd

```diff
@@ -0,0 +1,159 @@
+# Class inheriting HttpRouter for handling file serving requests
+extends HttpRouter
+class_name HttpFileRouter
+
+# Full path to the folder which will be exposed to web
+var path: String = ""
+
+# Full path to the fallback page which will be served if the requested file was not found
+var fallback_page: String = ""
+
+# An ordered list of extensions that will be checked 
+# if no file extension is provided by the request
+var extensions: PoolStringArray = ["html"]
+
+# A list of extensions that will be excluded if requested
+var exclude_extensions: PoolStringArray = []
+
+# Creates an HttpFileRouter intance
+#
+# #### Parameters
+# - path: Full path to the folder which will be exposed to web
+# - options: Optional Dictionary of options which can be configured.
+# 	- fallback_page: Full path to the fallback page which will be served if the requested file was not found
+#	- extensions: A list of extensions that will be checked if no file extension is provided by the request
+# 	- exclude_extensions: A list of extensions that will be excluded if requested
+func _init(
+	path: String, 
+	options: Dictionary = { 
+		fallback_page = fallback_page, 
+		extensions = extensions, 
+		exclude_extensions = exclude_extensions
+	}
+	) -> void:
+	self.path = path
+	self.fallback_page = options.get("fallback_page", "")
+	self.extensions = options.get("extensions", [])
+	self.exclude_extensions = options.get("exclude_extensions", [])
+
+# Handle a GET request
+func handle_get(request: HttpRequest, response: HttpResponse) -> void:
+	var serving_path: String = path + request.path
+	var file_exists: bool = _file_exists(serving_path)
+
+	if request.path.get_extension() == "" and not file_exists:
+		for extension in extensions:
+			serving_path = path + request.path + "." + extension
+			file_exists = _file_exists(serving_path)
+			if file_exists: 
+				break
+
+	# GDScript must be excluded, unless it is used as a preprocessor (php-like)
+	if (file_exists and not serving_path.get_extension() in ["gd"] + Array(exclude_extensions)):
+		response.send_raw(
+			200, 
+			_serve_file(serving_path), 
+			_get_mime(serving_path.get_extension())
+			)
+	else:
+		if fallback_page != "":
+			response.send_raw(404, _serve_file(fallback_page), _get_mime(fallback_page.get_extension()))
+		else:
+			response.send_raw(404)
+
+# (Internal function)
+# Reads a file as text
+#
+# #### Parameters
+# - file_path: Full path to the file
+func _serve_file(file_path: String) -> PoolByteArray:
+	var content: PoolByteArray = []
+	var file = File.new()
+	var file_opened: bool = not bool(file.open(file_path, File.READ))
```

Comment:
I wasn't planning on explaining why a file didn't open, so I did not think of returning an error.
But we can do this for sure if you want.
Instead of using a match (since matching each Godot error is something that I'd personally avoid) we could do something like
```
if file.open(file_path, File.READ):
  # error != OK --> error > 0 ( == TRUE)
  content = ("Couldn't serve file, ERROR = %s" % error).to_ascii()
else:
  # error = OK --> error =  0 ( == FALSE )
 ( ... )
```
We won't have a custom message for each single error, but at least we can print out the error code (which is a Global erorr in GDScript)
What do you think?

Reply:
That's okay I think.